### PR TITLE
Sliver Constrained Cross Axis

### PR DESCRIPTION
**Pull request by thkim1011**
Reimplements what we reverted here: #125233.

### CODE REVIEW

File: packages/flutter/lib/src/rendering/proxy_sliver.dart [hunk 1]
```diff
@@ -414,3 +415,43 @@ class RenderSliverAnimatedOpacity extends RenderProxySliver with RenderAnimatedO
...
+  @override
+  void performLayout() {
+    assert(child != null);
+    assert(maxExtent >= 0.0);
+    child!.layout(
+      constraints.copyWith(crossAxisExtent: min(_maxExtent, constraints.crossAxisExtent)),
+      parentUsesSize: true,
+    );
+    final SliverGeometry childLayoutGeometry = child!.geometry!;
+    geometry = childLayoutGeometry.copyWith(crossAxisExtent: min(_maxExtent, constraints.crossAxisExtent));
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/proxy_sliver.dart:455]: This is different from the original PR, can you say why?

File: packages/flutter/lib/src/rendering/sliver.dart [hunk 2]
```diff
@@ -719,6 +750,21 @@ class SliverGeometry with Diagnosticable {
   ///  * [RenderViewport.cacheExtent] for a description of a viewport's cache area.
   final double cacheExtent;
 
+  /// The amount of space allocated to the cross axis.
+  ///
+  /// This value should equal the [crossAxisExtent] of the sliver. If null, then
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver.dart:755]: Should it? Most cases this will be null. It is only relevant right now for the constrained cross axis sliver.

File: packages/flutter/lib/src/rendering/proxy_sliver.dart (hunk 1 above)
Comment (thkim1011) [reply packages/flutter/lib/src/rendering/proxy_sliver.dart:455]: I don't think this is different? Essentially, `SliverConstrainedCrossAxis` widgets have different `crossAxisGeometry` than the provided constraints, so we need to set it explicitly.

File: packages/flutter/lib/src/rendering/sliver.dart (hunk 2 above)
Comment (thkim1011) [reply packages/flutter/lib/src/rendering/sliver.dart:755]: You're right, I'll rephrase this to say that it will be null unless it is different from `SliverConstraints.crossAxisExtent`.

File: packages/flutter/lib/src/widgets/sliver.dart [hunk 3]
```diff
@@ -1365,3 +1365,100 @@ class KeepAlive extends ParentDataWidget<KeepAliveParentDataMixin> {
...
+      final AbstractNode? targetParent = renderObject.parent;
+      if (targetParent is RenderObject) {
+        targetParent.markNeedsLayout();
+      }
+
+    }
+  }
+
+  @override
+  Type get debugTypicalAncestorWidgetClass => Widget;
```
[the remark this answers is not in the document]
Comment (thkim1011) [reply packages/flutter/lib/src/widgets/sliver.dart:1439]: `SliverCrossAxisGroup` has yet to be defined, but I will change it in the `SliverCrossAxisGroup` PR.

File: packages/flutter/lib/src/rendering/proxy_sliver.dart (hunk 1 above)
Comment (thkim1011) [reply packages/flutter/lib/src/rendering/proxy_sliver.dart:455]: Oh my bad, you meant the original `SliverConstrainedCrossAxis` PR. Yeah, it's necessary for the `SliverGeometry` change, and necessary for when we implement `SliverCrossAxisGroup` PR in the future.

File: packages/flutter/lib/src/rendering/sliver.dart [hunk 4]
```diff
@@ -957,6 +1003,22 @@ class SliverPhysicalParentData extends ParentData {
   /// top left visible corner of the sliver.
   Offset paintOffset = Offset.zero;
 
+  /// The [crossAxisFlex] factor to use for this sliver child.
+  ///
+  /// If null or zero, the child is inflexible and determines its own size.
+  /// If non-zero, the amount of space the child can occupy in the
```
[the remark this answers is not in the document]
Comment (thkim1011) [reply packages/flutter/lib/src/rendering/sliver.dart]: Yes, I removed the part about the non-zero. Also, I changed `crossAxisFlex` to not have a default value to make this consistent with `FlexParentData`.

File: packages/flutter/lib/src/rendering/sliver.dart [hunk 5]
```diff
@@ -957,6 +1003,22 @@ class SliverPhysicalParentData extends ParentData {
...
+  /// cross axis is determined by dividing the free space (after placing the
+  /// inflexible children) according to the flex factors of the flexible
+  /// children.
+  ///
+  /// This value is only used by the [SliverCrossAxisGroup] widget to determine
```
[the remark this answers is not in the document]
Comment (thkim1011) [reply packages/flutter/lib/src/rendering/sliver.dart]: Removed!

File: packages/flutter/lib/src/rendering/sliver.dart [hunk 6]
```diff
@@ -957,6 +1002,11 @@ class SliverPhysicalParentData extends ParentData {
   /// top left visible corner of the sliver.
   Offset paintOffset = Offset.zero;
 
+  /// The [crossAxisFlex] factor to use for this sliver child.
+  ///
+  /// If null or zero, the child is inflexible and determines its own size in the cross axis.
+  int? crossAxisFlex;
```
Comment (Piinks) [inline packages/flutter/lib/src/rendering/sliver.dart:1008]: Have you tried this with the SliverGroup? I think I recall there was an issue with wanting to flex normal slivers along with SliverCross..Expanded. How will the group determine that a normal sliver, say SliverList, should flex if it's flex value is null?

Comment (Piinks) [reply packages/flutter/lib/src/rendering/sliver.dart:1008]: I think this is the only item left here to resolve. :)

Comment (thkim1011) [reply packages/flutter/lib/src/rendering/sliver.dart:1008]: Yep I've tried this with SliverGroup and it works well. Essentially `SliverGroup`'s `setupParentData` method will create children `parentData` with `crossAxisFlex` of 1. However, the `ParentDataWidget`s for the `SliverCrossAxisConstrained` will set this to 0, for the constrained widgets.

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
